Add Annoy SideInput

fallonchen opened the pull request:
* Use annoy4s for building index
* Use annoy-java for reading

nevillelyh commented:
Do you have any benchmark numbers? Paste here?

fallonchen commented:
I have a basic one based on 18M w2v vectors; I'm redoing my benchmark because I think it's weird that annoy4s lookups are so much slower:

**Annoy4s lookup benchmark (getNByVector)**
Average time to get item: 2.5939111417042955E-4
Total time to get items: 2.5939111417042955E-4
Total items: 18707111

** annoy-java lookup getNearest **
Average time to get item: 2.0688557659026075E-7
Total time to get items: 2.0688557659026075E-7

ravwojdyla commented on scio-examples/src/main/scala/com/spotify/scio/examples/extra/AnnoyExample.scala:
```diff
@@ -0,0 +1,59 @@
+import com.spotify.scio.ContextAndArgs
+import com.spotify.scio.extra.annoy._
+import com.spotify.scio.values.SCollection
+/**
```
new line

ravwojdyla commented on scio-examples/src/main/scala/com/spotify/scio/examples/extra/AnnoyExample.scala:
```diff
@@ -0,0 +1,59 @@
+    val nTrees = 10
+
+    // open annoy file from scio context
+    //val path = args("input")
```
What is this comment for? if it's another example maybe there should be two examples?

ravwojdyla replied:
Alright - i see, maybe just have the `args` version here, and the rest could be documented in tests?

fallonchen replied:
It's meant to be another example, but the code for accessing the data loaded would be similar. Would it be better to put a switch in the command line arguments?

ravwojdyla replied:
I think it probably better to keep this as close to "real job" as possible, and real job would not create dummy data. I think i would have just one example but close to "real job".

ravwojdyla commented on scio-extra/src/it/scala/annoy/AnnoyIT.scala:
```diff
@@ -0,0 +1,79 @@
+          .flatMap((xs, si) => xs.map(x => si(p2).getItemVector(x))).toSCollection
+        s should containInAnyOrder(sideData.map(_._2))
+      } finally {
+        val files = FileSystems
```
Is this gonna throw in case of failure? is `files` needed for anything?

fallonchen replied:
This line was missing: FileSystems.delete(files.asJava); will add that in.

ravwojdyla commented on scio-extra/src/it/scala/annoy/AnnoyIT.scala:
```diff
@@ -0,0 +1,79 @@
+    runWithContext { sc =>
+      FileSystems.setDefaultPipelineOptions(sc.options)
+      val tempLocation = ItUtils.gcpTempLocation("annoy-it")
+      val path = tempLocation + "/annoy.tree"
```
`s""`?

ravwojdyla commented on scio-extra/src/main/scala/com/spotify/scio/extra/annoy/AnnoyUri.scala:
```diff
@@ -0,0 +1,106 @@
+    val tempFile = Files.createTempDirectory("annoy").resolve("data")
+    w.build()
+    w.save(tempFile.toString)
+    w.free()
```
is this sth that should be in `finally`?

ravwojdyla commented on scio-extra/src/main/scala/com/spotify/scio/extra/annoy/AnnoyUri.scala:
```diff
@@ -0,0 +1,106 @@
+  override private[annoy] def saveAndClose(w: AnnoyWriter): Unit = {
+    w.build
+    w.save(path.toString)
+    w.free()
```
Is this sth that should be in `finally`?

fallonchen replied:
Like this `try { w.build(); w.save(path.toString) } finally { w.free() }`?

ravwojdyla commented on scio-extra/src/main/scala/com/spotify/scio/extra/annoy/package.scala:
```diff
@@ -0,0 +1,214 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
```
new line

ravwojdyla commented on scio-extra/src/main/scala/com/spotify/scio/extra/annoy/package.scala:
```diff
@@ -0,0 +1,214 @@
+
+import scala.collection.JavaConverters._
+
+
```
- new line

ravwojdyla commented on scio-extra/src/main/scala/com/spotify/scio/extra/annoy/package.scala:
```diff
@@ -0,0 +1,214 @@
+  case object Angular extends ScioAnnoyMetric
+  case object Euclidean extends ScioAnnoyMetric
+
+  val logger = LoggerFactory.getLogger(this.getClass)
```
`private`?

ravwojdyla commented on scio-extra/src/main/scala/com/spotify/scio/extra/annoy/package.scala:
```diff
@@ -0,0 +1,214 @@
+    def getItemVector(i: Int): Array[Float] = index.getItemVector(i)
+    def getItemVector(i: Int, v: Array[Float]): Unit = index.getItemVector(i, v)
+    def getNearest(v: Array[Float], maxNumResults: Int): Array[Int] = {
+      index.getNearest(v, maxNumResults).asScala.map(Int.box(_).toInt).toArray[Int]
```
`Integer2int`?

ravwojdyla commented on scio-extra/src/main/scala/com/spotify/scio/extra/annoy/package.scala:
```diff
@@ -0,0 +1,214 @@
+    def annoySideInput(metric: ScioAnnoyMetric, dim: Int, nTrees: Int, path: String)
+    : SideInput[AnnoyReader] = {
+      val uri = AnnoyUri(path, self.options)
+      val view = self.parallelize(Seq(uri)).applyInternal(View.asSingleton())
```
Any reason why not `asSingletonSideInput`?

ravwojdyla commented on scio-extra/src/main/scala/com/spotify/scio/extra/annoy/package.scala:
```diff
@@ -0,0 +1,214 @@
+    }
+  }
+
+  private class AnnoySideInput(val view: PCollectionView[AnnoyUri],
```
I assume there is no need to validate the constructor parameters or any other place earlier when we use it?

fallonchen replied:
Do you mean sanity checking that the values of metric, dim and nTrees fall in an appropriate range? I was relying on the user to be familiar enough with annoy to provide good values for that.

ravwojdyla replied:
I mean simple checks like dim larger than 0, there. There also was some hidden feature in annoy build, where if you set something to `-1` it will find optimal value etc.

ravwojdyla replied:
https://github.com/spotify/annoy/issues/216 and i think you could also set the number of trees to`-1` (https://github.com/spotify/annoy/commit/ef4051a8a15684222ce46208ff1a30b8b8a3ee35)

ravwojdyla commented on scio-extra/src/main/scala/com/spotify/scio/extra/annoy/package.scala:
```diff
@@ -0,0 +1,214 @@
+              val kv = it.next()
+              annoyIndex.addItem(kv._1, kv._2)
+            }
+            uri.saveAndClose(annoyIndex)
```
Is there a need for this to be in `finally`?

ravwojdyla commented on scio-extra/src/main/scala/com/spotify/scio/extra/annoy/package.scala:
```diff
@@ -0,0 +1,214 @@
+            val annoyIndex = new AnnoyWriter(metric, dim, nTrees)
+            val it = xs.iterator
+            while (it.hasNext) {
+              val kv = it.next()
```
`val (k, v)` would be tiny bit better looking :)

ravwojdyla commented on scio-extra/src/main/scala/com/spotify/scio/extra/annoy/package.scala:
```diff
@@ -0,0 +1,214 @@
+      }
+    }
+
+    /**
```
Comment `*` formating

ravwojdyla commented on scio-extra/src/main/scala/com/spotify/scio/extra/annoy/package.scala:
```diff
@@ -0,0 +1,214 @@
+    def asAnnoy(metric: ScioAnnoyMetric, dim: Int, nTrees: Int)
+    : SCollection[AnnoyUri] = {
+      val uuid = UUID.randomUUID()
+      val path = self.context.options.getTempLocation + s"/annoy-build-$uuid"
```
probably the whole thing could be in `s""` also is there a need to check if `getTempLocation` ends with `/`?

ravwojdyla commented on scio-extra/src/test/scala/com/spotify/scio/extra/annoy/AnnoyTest.scala:
```diff
@@ -0,0 +1,84 @@
+  }
+
+
+  "SCollection" should "support .asAnnoy with specified local file" in {
```
`"SCollection"` -> `it`

ravwojdyla commented on scio-extra/src/main/scala/com/spotify/scio/extra/annoy/package.scala:
```diff
@@ -0,0 +1,214 @@
+
+  implicit class AnnoyPairSCollection(val self: SCollection[(Int, Array[Float])]) {
+    /**
+      * Write the key-value pairs of this SCollection as an Annoy file to a specific location.
```
`*` formatting

ravwojdyla commented on scio-extra/src/main/scala/com/spotify/scio/extra/annoy/package.scala:
```diff
@@ -0,0 +1,214 @@
+
+  /** Enhanced version of [[ScioContext]] with Annoy methods */
+  implicit class AnnoyScioContext(val self: ScioContext) extends AnyVal {
+    /**
```
`*` formatting

ravwojdyla commented on scio-extra/src/main/scala/com/spotify/scio/extra/annoy/package.scala:
```diff
@@ -0,0 +1,214 @@
+  }
+
+  /**
+    * Enhanced version of [[com.spotify.scio.values.SCollection SCollection]] with Annoy methods
```
And comments `*` again

ravwojdyla commented on scio-extra/src/main/scala/com/spotify/scio/extra/annoy/package.scala:
```diff
@@ -0,0 +1,214 @@
+  /**
+   *  Wraps annoy-java
+   */
+  class AnnoyReader(metric: ScioAnnoyMetric, dim: Int, nTrees: Int, path: String) {
```
Should `AnnoyReader` have public CTOR? If so probably should have more doc.

ravwojdyla commented on scio-extra/src/main/scala/com/spotify/scio/extra/annoy/package.scala:
```diff
@@ -0,0 +1,214 @@
+    */
+  implicit class AnnoySCollection(val self: SCollection[AnnoyUri]) extends AnyVal {
+    def asAnnoySideInput(metric: ScioAnnoyMetric, dim: Int, nTrees: Int)
+    : SideInput[AnnoyReader] = {
```
RTP still can be in previous line

ravwojdyla commented on scio-extra/src/main/scala/com/spotify/scio/extra/annoy/package.scala:
```diff
@@ -0,0 +1,214 @@
+     * @return SideInput[AnnoyReader]
+     */
+    def asAnnoySideInput(metric: ScioAnnoyMetric, dim: Int, nTrees: Int)
+    : SideInput[AnnoyReader] = {
```
RTP still can be in previous line

ravwojdyla commented on scio-extra/src/main/scala/com/spotify/scio/extra/annoy/package.scala:
```diff
@@ -0,0 +1,214 @@
+      *   @return A Singleton SCollection containing the [[AnnoyUri]] of the saved files
+      */
+    def asAnnoy(metric: ScioAnnoyMetric, dim: Int, nTrees: Int)
+    : SCollection[AnnoyUri] = {
```
RTP still can be in previous line

ravwojdyla commented on scio-extra/src/main/scala/com/spotify/scio/extra/annoy/AnnoyUri.scala:
```diff
@@ -0,0 +1,106 @@
+  override private[annoy] def getReader(metric: ScioAnnoyMetric, dim: Int, nTrees: Int): AnnoyReader
+  = new AnnoyReader(metric, dim, nTrees, path)
+  override private[annoy] def saveAndClose(w: AnnoyWriter): Unit = {
+    w.build
```
+ `()`

fallonchen commented:
Using ~18M 40 dimension vectors, 10 trees, 10 search results, search_k set to (numtrees * num results):

////////////// annoy java //////////////////////
Total load time (ms): 4075.186
Total lookup time (ms) for 18707111 items: 3784199.757
Average lookup time (ms) for 18707111 items: 0.202

////////////// annoy4s  //////////////////////
Total load time (ms): 2354.253
Total lookup time (ms) for 18707111 items: 4.273816706563E9
Average lookup time (ms) for 18707111 items: 228.4594722596664
////////////// annoy4s  //////////////////////

fallonchen commented:
@ravwojdyla Made changes according to code review. Please take a look when you have time! 

tl;dr on changes:
* formatting cleanup
* more documentation on AnnoyReader CTOR and other public facing functions
* removed unused member nTrees in AnnoyReader class
* moved index cleanup into try / finally
* split example into 2 examples, one for loading side inputs and one for creating annoy index files

ravwojdyla commented on scio-examples/src/main/scala/com/spotify/scio/examples/extra/AnnoyIndexSaveExample.scala:
```diff
@@ -0,0 +1,43 @@
+import com.spotify.scio.values.SCollection
+
+/**
+ * sbt "project scio-examples" "runMain com.spotify.scio.examples.extra.AnnoyIndexSaveExample \
```
Afair other examples don't use `\`, as much as yours is more "correct", would be nice to stick with one convention?

ravwojdyla commented on scio-examples/src/main/scala/com/spotify/scio/examples/extra/AnnoySideInputExample.scala:
```diff
@@ -0,0 +1,55 @@
+import com.spotify.scio.extra.annoy._
+import com.spotify.scio.values.SCollection
+/**
+ * sbt "project scio-examples" "runMain com.spotify.scio.examples.extra.AnnoySideInputExample \
```
Same as above.

ravwojdyla commented on scio-extra/src/test/scala/com/spotify/scio/extra/annoy/AnnoyTest.scala:
```diff
@@ -0,0 +1,83 @@
+    new File(path).delete()
+  }
+
+  "SCollection" should "support .asAnnoy with specified local file" in {
```
`it`?

ravwojdyla commented:
@yonromai fyi

yonromai commented:
@fallonchen  A few comments: 

- perhaps I'm missing something, but is there a reason for using `annoy-java` at all? `annoy-java` hasn't been updated in a while (unlike the original annoy cpp code base). Also, it seems `annoy4s` supports reading indexes too and you have guarantees that the annoy implementation used to read the tree is the same than the one that was used to create it

- By any chance, did you compare the quality of the neighbors returned by `annoy-java` with `annoy-4s`? I wouldn't be surprise if there is a significant difference. 

- FYI, a while back I worked on a JNI implementation of annoy java. I seems to be very similar to `annoy4s` with a few minor differences: 

  - The annoy cpp code is build by maven and bundle on the jar on the fly (unlike annoy4s which has it in the code). This should make it easier to follow the annoy-cpp updates
  - There is a string API which uses sparkey indexes transparently to make string<->int keys
  - The code is there: https://ghe.spotify.net/romain/annoy-java. I published the jar on mvn central (it's used by inferno in prod). I have on my roadmap to open source it but I need to get in touch with FOSS to do that .

fallonchen commented:
@yonromai 

> perhaps I'm missing something, but is there a reason for using annoy-java at all? annoy-java hasn't been updated in a while (unlike the original annoy cpp code base). Also, it seems annoy4s supports reading indexes too and you have guarantees that the annoy implementation used to read the tree is the same than the one that was used to create it

With annoy4s, we have to explicitly free the memory allocated to the index. We talked about wrapping this in an explicit `.close()` method and having the user remember to call it each time, but it seemed like a better solution to keep things clean, use annoy-java (assuming it performed equally well as annoy4s a reader) and have GC handle the index cleanup.

> By any chance, did you compare the quality of the neighbors returned by annoy-java with annoy-4s? I wouldn't be surprise if there is a significant difference.

I didn't, but I do think that's a good idea. I'll look into it :)

> FYI, a while back I worked on a JNI implementation of annoy java. I seems to be very similar to annoy4s with a few minor differences:

Cool, will check this out!

yonromai commented:
Great to know! In my JNI implementation I used `AutoClosable` ([code](https://ghe.spotify.net/romain/annoy-java/blob/master/src/main/java/com/spotify/annoy/jni/base/AnnoyIndex.java#L29))  but I'm wondering if it actually releases memory correctly :)
Keep me posted!
